Add CombinedRocFilter and roc-intersect-mode to vcfeval

rickymagner opened the pull request:
This PR adds a flag to `vcfeval` called `--roc-intersect-mode` which toggles whether the ROC filtering conditions should be applied together ("AND") as opposed to in parallel separately (current behavior). 

## The Feature

An example command is the following:
```
rtg vcfeval -b base.vcf -c call.vcf -t ref -o out_dir \
    --roc-intersect-mode \
    --roc-regions test-reg=test1.bed \
    --roc-regions test-reg2=test2.bed \
    --roc-subset snp,non-snp,het \
    --roc-expr "stuff=INFO.TEST>5"
```

Normally, this command would produce 6 `_roc.tsv.gz` files for each of the different categories. However, when the flag `--roc-intersect-mode` is included you will get a ROC output for each possible combination of the filter conditions. In the `subset` category, two subcategories are defined internally to separate variant conditions (SNP, etc) from genotype conditions (Het, etc). In particular, each provided region, subset, and expression would be combined (along with a default "no filter" in each category) to provide (2+1 regions) x (2+1 variant types) x (1+1 genotypes) x (1+1 expressions) = 36 files in this case. 

So for example, the above command would provide ROC file outputs for e.g. SNPs in test1.bed (`test-reg+snp_roc.tsv.gz`), non-SNP hets in test2.bed with`INFO.TEST>5` (`stuff+test-reg2+het+non_snp_roc.tsv.gz`), and so on.

## The Implementation

The code attempts to minimally change existing code, and utilizes the framework for applying ROC filters. It introduces a new class called `CombinedRocFilter` which is created from a list of `RocFilter`s and holds them together. This object is itself a `RocFilter`, and it `accept`s a site exactly when it passes the `accept` criteria for all of the filters it holds, thus applying the intersection of all conditions it holds. Because it is a `RocFilter`, it can be dropped in place for where `RocFilter`s were previously used for generating the output stat files.

A little bit of edge case reasoning is done to handle producing outputs where an arbitrary subset of filters is applied (e.g. just some of the different categories, like variant type and region, but not the others), especially the case where "no filter" is applied for global stats and backwards compatibility in the usual output. In the case the user does not toggle this flag, no `CombinedRocFilter`s should be created, and the tool should work the same as before, applying each filter separately and generating its own 1 output file per filter.

Please let me know if you have any questions, comments, or suggestions to improve the code. Thank you!

rickymagner committed:
vcfeval: Add --roc-cross-join mode to vcfeval, vcf2rocplot, cnveval

When enabled, produces ROC files for every possible intersection of
user-specified ROC criteria groups. Groups:

- Any --roc-region values form one group
- Any --roc-expr values form one group
- Any --roc-subset values are either a "variant type" group (e.g. snp /
  mnp / indel), or a "genotype" group (e.g. het / hom)

Each resulting cross-join ROC will have rescaling enabled only if all filters
on a combination have rescaling enabled. E.g. using the flags:

  --roc-region my_bed.bed --roc-subset het,hom

will output two ROC files, one for heterozygous variants within my_bed.bed
regions, the other for homozygous variants within my_bed.bed regions.

Lenbok commented on src/main/java/com/rtg/vcf/eval/CombinedRocFilter.java:
```diff
@@ -0,0 +1,39 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class CombinedRocFilter extends RocFilter {
```
This should probably implement `requiresGt` to return true if any of the delegate filters require the GT.

Lenbok replied:
It would also be great if you could add a unit test for this class.

rickymagner replied:
Good call, just added a simple implementation. I'll work on getting the unit test for this class.

Lenbok commented on src/main/java/com/rtg/vcf/eval/CombinedRocFilter.java:
```diff
@@ -0,0 +1,39 @@
+    private final ArrayList<RocFilter> mRocFilters = new ArrayList<>();
+
+    /**
+     * Create a CombinedRocFilter which incorporates multiple RocFilters into one filter
```
A little more informative description.
```suggestion
     * Create a CombinedRocFilter which accepts only records that are accepted by all delegate RocFilters.
```

rickymagner replied:
Good improvement; added.

Lenbok commented on src/main/java/com/rtg/vcf/eval/VcfEvalCli.java:
```diff
@@ -92,6 +93,7 @@ public class VcfEvalCli extends ParamsCli<VcfEvalParams> {
   protected static final String ROC_SUBSET = "roc-subset";
   protected static final String ROC_EXPR = "roc-expr";
   protected static final String ROC_REGIONS = "roc-regions";
+  protected static final String ROC_INTERSECT_MODE = "roc-intersect-mode";
```
I'm not a fan of the flag name. Perhaps:
```suggestion
  protected static final String ROC_CROSS_JOIN = "roc-cross-join";
```

rickymagner replied:
Updated the name

Lenbok commented on src/main/java/com/rtg/vcf/eval/VcfEvalCli.java:
```diff
@@ -215,6 +217,7 @@ public static <T extends RocFilterProxy> void registerVcfRocFlags(CFlags flags,
     flags.registerOptional(ROC_SUBSET, rocfilterclass, CommonFlags.STRING, "output ROC file for preset variant subset").setMaxCount(Integer.MAX_VALUE).enableCsv().setCategory(REPORTING);
     flags.registerOptional(ROC_EXPR, String.class, CommonFlags.STRING, "output ROC file for variants matching custom JavaScript expression. Use the form <LABEL>=<EXPRESSION>").setMaxCount(Integer.MAX_VALUE).setCategory(REPORTING);
     flags.registerOptional(ROC_REGIONS, String.class, CommonFlags.STRING, "output ROC file for variants overlapping custom regions supplied in BED file. Use the form <LABEL>=<FILENAME>").setMaxCount(Integer.MAX_VALUE).setCategory(REPORTING);
+    flags.registerOptional(ROC_INTERSECT_MODE, "toggle producing ROC files for each intersection of ROC_SUBSET, ROC_EXPR, and ROC_REGIONS values provided").setCategory(REPORTING);
```
```suggestion
    flags.registerOptional(ROC_INTERSECT_MODE, "output ROC files for each intersection of --" + ROC_SUBSET + ", --" + ROC_EXPR + ", and --" + ROC_REGIONS + " values provided").setCategory(REPORTING);
```

rickymagner replied:
Nice, I didn't see the constants for these flag values. Just updated this string.

Lenbok commented on src/main/java/com/rtg/vcf/eval/VcfEvalCli.java:
```diff
@@ -400,29 +403,84 @@ protected VcfEvalParams makeParams() throws IOException {
+        rocExprFilters.add(new ExpressionRocFilter(e[0], e[1]));
       }
     }
+    rocFiltersTypeMap.put("rocExprFilters", rocExprFilters);
```
Can we use the flag name constants as the keys in this map? I'm hesitant to use copy-pasted strings that are susceptible to typos. e.g.:
```suggestion
    rocFiltersTypeMap.put(ROC_EXPR, rocExprFilters);
```

rickymagner replied:
This is a good point. I updated it for `ROC_EXPR` and `ROC_REGIONS`, but unsure what to do about `rocGenotypeFilters` and `rocVariantFilters` since they didn't have analogous constants.

Lenbok commented on src/main/java/com/rtg/vcf/eval/VcfEvalCli.java:
```diff
@@ -400,29 +403,84 @@ protected VcfEvalParams makeParams() throws IOException {
+    }
+    rocFiltersTypeMap.put("rocRegionsFilters", rocRegionsFilters);
+
+    // Split subset filters into genotype or variant flavors for intersecting in ROC_INTERSECT_MODE
```
Yeah, this splitting of the `ROC_SUBSET` members into two subgroups is perhaps not obvious to the end user. Will think about this a little more.

rickymagner replied:
I think if you want to mix and match ROC stats across different types of subsets, it's natural to also split variant type and zygosity into separate dimensions for the subsetting. There will still be outputs for the larger categories, e.g. just Hets, just SNPs, etc. but also for the cross between them.

Lenbok reviewed (changes requested):
This is looking pretty nice, thanks!

As well as the specific unit test suggested, can you please add a small integration test in `VcfEvalNanoTest` that exercises this new option. Have a look at some of the examples there, essentially you give some small input files and check the expected outputs are present (most of the provided tests are checking output VCFs, but e.g. see how `testNanoSmallAlt` gives a parameter containing the names of the output files to check)

rickymagner commented:
Just pushed some minor changes to address some of the documentation changes requested above. I'll work on trying to get the tests worked out sometime soon, hopefully this week. Thanks!

rickymagner commented:
Hi @Lenbok, I just added some tests as suggested for the new feature (or at least attempted to). Do you know how I can try running them to make sure they're correct? I'd also appreciate if you took a look at them to see if you're satisfied with the case coverage at some point. Thanks!

Lenbok commented:
Hi @rickymagner, from the command line you can run all the unit tests via `ant runalltests`, and run a specific test class using (e.g) `ant runtest -Dtest.class=com.rtg.vcf.eval.VcfEvalNanoTest`. Depending on what IDE you use, you can probably run individual test classes/tests directly from there.

(Aside, I also just noticed you are using four space indentation, whereas we use two spaces)

Lenbok commented on src/test/java/com/rtg/vcf/eval/CombinedRocFilterTest.java:
```diff
@@ -0,0 +1,133 @@
+    }
+
+    public void testAll() {
+        final RocFilter f = CombinedRocFilter(Arrays.asList(RocFilter.ALL), false);
```
```suggestion
        final RocFilter f = new CombinedRocFilter(Arrays.asList(RocFilter.ALL), false);
```
(and in the similar cases below)

Lenbok commented:
You will also see there are some test failures during to spell-checking the command line help - it looks like you need to add `roc-expr` to the exclusions list in `src/test/resources/com/rtg/util/cli/spell.sensitive` (there may be others needed)

rickymagner commented:
Hi Len, I just addressed a few of your comments:
1. I switched from 4 spaces to 2 spaces for indent in the relevant files.
2. Fixed some of the failing filter tests.
3. Added a spellcheck exclusion for `roc-expr` which helped with the tests.

I'm still seeing two tests fail when testing locally:
1. Something involving the usage server, which I imagine is just specific to my machine and maybe not a general issue?
```
     [java] 1) test(com.rtg.usage.UsageServerCliTest)
     [java] junit.framework.AssertionFailedError: Checking usage configuration.
     [java] Clients will not be able to connect until RTG_USAGE_HOST has been set. Please consult the user manual section "Advanced Installation Configuration".
     [java] Error: Could not create socket on port 3283. Address already in use
     [java]  expected:<0> but was:<1>
     [java] 	at junit.framework.Assert.fail(Assert.java:57)
     [java] 	at junit.framework.Assert.failNotEquals(Assert.java:329)
     [java] 	at junit.framework.Assert.assertEquals(Assert.java:78)
     [java] 	at junit.framework.Assert.assertEquals(Assert.java:234)
     [java] 	at junit.framework.TestCase.assertEquals(TestCase.java:401)
     [java] 	at com.rtg.usage.UsageServerCliTest.test(UsageServerCliTest.java:98)
     [java] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
     [java] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
     [java] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
     [java] 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
     [java] 	at junit.framework.TestCase.runTest(TestCase.java:176)
     [java] 	at junit.framework.TestCase.runBare(TestCase.java:141)
     [java] 	at junit.framework.TestResult$1.protect(TestResult.java:122)
     [java] 	at junit.framework.TestResult.runProtected(TestResult.java:142)
     [java] 	at junit.framework.TestResult.run(TestResult.java:125)
     [java] 	at junit.framework.TestCase.run(TestCase.java:129)
     [java] 	at junit.framework.TestSuite.runTest(TestSuite.java:252)
     [java] 	at junit.framework.TestSuite.run(TestSuite.java:247)
     [java] 	at org.junit.internal.runners.JUnit38ClassRunner.run(JUnit38ClassRunner.java:86)
     [java] 	at org.junit.runners.Suite.runChild(Suite.java:128)
     [java] 	at org.junit.runners.Suite.runChild(Suite.java:27)
     [java] 	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
     [java] 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
     [java] 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
     [java] 	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
     [java] 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
     [java] 	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
     [java] 	at org.junit.runners.Suite.runChild(Suite.java:128)
     [java] 	at org.junit.runners.Suite.runChild(Suite.java:27)
     [java] 	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
     [java] 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
     [java] 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
     [java] 	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
     [java] 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
     [java] 	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
     [java] 	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
     [java] 	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
     [java] 	at org.junit.runner.JUnitCore.run(JUnitCore.java:105)
     [java] 	at org.junit.runner.JUnitCore.run(JUnitCore.java:94)
     [java] 	at RtgTestEntry.main(RtgTestEntry.java:175)
```

2. A problem with `VcfEvalNanoTest`. 
```
     [java] 2) testNanoRocCrossJoin(com.rtg.vcf.eval.VcfEvalNanoTest)
     [java] java.io.FileNotFoundException: /var/folders/w5/8rg1mj0j1rx9skg02s9s_r900000gp/T/unitTest7985174708251715361vcfeval-nano/output/snp_roc.tsv.gz (No such file or directory)
     [java] 	at java.base/java.io.FileInputStream.open0(Native Method)
     [java] 	at java.base/java.io.FileInputStream.open(FileInputStream.java:219)
     [java] 	at java.base/java.io.FileInputStream.<init>(FileInputStream.java:157)
     [java] 	at java.base/java.io.FileReader.<init>(FileReader.java:75)
     [java] 	at com.rtg.util.io.FileUtils.fileToString(FileUtils.java:639)
     [java] 	at com.rtg.launcher.AbstractEndToEndTest.checkResultFile(AbstractEndToEndTest.java:91)
     [java] 	at com.rtg.launcher.AbstractEndToEndTest.lambda$endToEnd$0(AbstractEndToEndTest.java:56)
     [java] 	at com.rtg.launcher.AbstractEndToEndTest.checkResults(AbstractEndToEndTest.java:85)
     [java] 	at com.rtg.vcf.eval.AbstractVcfEvalTest.endToEnd(AbstractVcfEvalTest.java:73)
     [java] 	at com.rtg.launcher.AbstractEndToEndTest.endToEnd(AbstractEndToEndTest.java:54)
     [java] 	at com.rtg.launcher.AbstractEndToEndTest.endToEnd(AbstractEndToEndTest.java:50)
     [java] 	at com.rtg.vcf.eval.VcfEvalNanoTest.testNanoRocCrossJoin(VcfEvalNanoTest.java:164)
     [java] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
     [java] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
     [java] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
     [java] 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
     [java] 	at junit.framework.TestCase.runTest(TestCase.java:176)
     [java] 	at junit.framework.TestCase.runBare(TestCase.java:141)
     [java] 	at junit.framework.TestResult$1.protect(TestResult.java:122)
     [java] 	at junit.framework.TestResult.runProtected(TestResult.java:142)
     [java] 	at junit.framework.TestResult.run(TestResult.java:125)
     [java] 	at junit.framework.TestCase.run(TestCase.java:129)
     [java] 	at junit.framework.TestSuite.runTest(TestSuite.java:252)
     [java] 	at junit.framework.TestSuite.run(TestSuite.java:247)
     [java] 	at org.junit.internal.runners.JUnit38ClassRunner.run(JUnit38ClassRunner.java:86)
     [java] 	at org.junit.runners.Suite.runChild(Suite.java:128)
     [java] 	at org.junit.runners.Suite.runChild(Suite.java:27)
     [java] 	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
     [java] 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
     [java] 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
     [java] 	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
     [java] 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
     [java] 	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
     [java] 	at org.junit.runners.Suite.runChild(Suite.java:128)
     [java] 	at org.junit.runners.Suite.runChild(Suite.java:27)
     [java] 	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
     [java] 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
     [java] 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
     [java] 	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
     [java] 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
     [java] 	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
     [java] 	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
     [java] 	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
     [java] 	at org.junit.runner.JUnitCore.run(JUnitCore.java:105)
     [java] 	at org.junit.runner.JUnitCore.run(JUnitCore.java:94)
     [java] 	at RtgTestEntry.main(RtgTestEntry.java:175)
```
Here it cannot compare the files properly. The way it's currently written, the command outputs `*_roc.tsv.gz` files which are compressed, but I committed non-compressed files. The issue is that two gzipped files can't be compared directly because there is often metadata like timestamps included in their headers. The way the testing infrastructure seems to be set up, it only allows comparing file strings directly to assert they are equal.

But even compressing my test output files doesn't seem to help. I get the same stacktrace saying it cannot find the right files.

What do you think is the best way to resolve this? I'm not so sure it's as much a matter about comparing .gz files (though suspect that might eventually be an issue if not this one) or rather I am not calling the function with the right paths.

Lenbok commented:
Hi @rickymagner,

Regarding the usage server test, I'm not sure what is causing that - the error message indicates something on your machine might already be listening on port 3283?

Regarding `VcfEvalNanoTest` though, the `endToEnd` test adds on the `-Z` flag for uncompressed output files, so you need to remove the `.gz` from the files you're listing that need to be checked. The next thing is that during regression results comparison, the files undergo simple sanitization to remove version numbers and command lines, but the expected outputs you've committed still have those present.

If you run the test with `ant runtest -Dtest.class=com.rtg.vcf.eval.VcfEvalNanoTest '-Dregression.update=.*'` the nanoregression will update those expected outputs (and you'll be able to see from git diff that it's just those header lines that change).

rickymagner commented:
Great, it looks like that worked to get all tests passing! Not sure what was listening on that port before, but after a restart all the tests passed for me. I just committed those small version updates.

Please let me know if there's anything else I can do to help this get merged! Thanks for all your help so far.

Lenbok commented:
Hi @rickymagner, sorry for being slow on the replies here. 

I've got just two hesitations:
- It divides the pre-canned subset filters into genotype vs variant filters, but that may not be obvious to the user (since they originate from the one `--roc-subset` flag). Possibly those pre-canned ones should be set via separate  `--roc-gt-subset` and `--roc-variant-subset` making it more obvious how the cross-join option comes up with its combinations.
- I'm considering whether a third, related alternative is useful, where rather than creating multiple outputs using the cross product, we output a single roc with the intersection (this was a user request here: https://groups.google.com/a/realtimegenomics.com/g/rtg-users/c/-FMLpcsi9ec/m/K0nM3kIDAwAJ). In which it would be nicer to just have a single flag that controls the mode, e.g. `--roc-subset-mode={default, intersection, cross-join}`

However, neither of those need to be addressed immediately, if we just denote the new flag in your PR as experimental by renaming it to `--Xroc-cross-join`. That way we can tidy the above issues up at a later date. If you're happy with this we can merge it in.

rickymagner commented:
Hi @Lenbok, thanks for the reply. I agree with your first point that it'd be nice to make it clear that we'd be considering SNP vs INDEL as an orthogonal choice to HET vs HOM, when they were previously considered on the same footing with that flag. Would you suggest actually replacing the `--roc-subset` flag entirely with those two? I'd be in favor of it from a technical standpoint, but worry about breaking backwards compatibility.

For the second point, I agree it could be a little easier to work with having one output ROC rather than a possibly large collection (and then having to remember the inputs to the command to grab them all up into one table, or at least that's what I planned to do). If I'm understanding correctly, the flag `--roc-subset-mode` would be new and have the behavior:
* `default`: the current behavior
* `intersection`: combine orthogonal filters together "multiplicatively" rather than the current "additive" behavior, i.e. snp + interval would just counts SNPs over that interval
* `cross-join`: the behavior of this PR (but maybe combined into just one ROC? How would that look? Some extra column(s) to designate the choice for that row?)

I think the second option sounds more appealing from the user perspective, although would require a bit more work. I'm OK with potentially trying to get a flag like that implemented if you agree since it shouldn't be much harder than what's already been done here. I'd prefer to try to get things done "right" before modifying the pipelines I want to include it in so those won't have to be changed so quickly after.

Lenbok commented:
For the first point I would definitely not want to break backward compatibility either. We don't necessarily *have* to do something about this, it's just less than optimal.

For the second you have the right interpretation. Although, now that I think of it, `intersection` is just a special case of `cross-join` where none of the "orthogonal filters" had more than one value selected. So maybe we don't need to leave room for three modes.

Lenbok commented on src/main/java/com/rtg/vcf/eval/VcfEvalCli.java:
```diff
@@ -400,29 +403,84 @@ protected VcfEvalParams makeParams() throws IOException {
+
+    final Set<RocFilter> rocFilters = new LinkedHashSet<>();
+    if (flags.isSet(ROC_CROSS_JOIN)) {
+      // Add decoy "ALL" filters to ensure following cartesian product isn't empty and hit every combination
+      rocFiltersTypeMap.get(ROC_EXPR).add(RocFilter.ALL);
+      rocFiltersTypeMap.get(ROC_REGIONS).add(RocFilter.ALL);
+      rocFiltersTypeMap.get("rocGenotypeFilters").add(RocFilter.ALL);
+      rocFiltersTypeMap.get("rocVariantFilters").add(RocFilter.ALL);
```
Perhaps we should only add these `RocFilter.ALL` entries to the maps if they are otherwise empty.

If the user gives `--roc-cross-join --roc-subset het,hom,snp` I think I would expect the following output files:
- weighted_roc.tsv (which we always produce)
- het+snp_roc.tsv
- hom+snp_roc.tsv

I don't think I would expect to see:
- snp_roc.tsv
- het_roc.tsv
- hom_roc.tsv

If the user gives `--roc-cross-join --roc-subset het,hom,snp --roc-expr "test=INFO.TEST>5" --roc-regions "mybed=mybed.bed"` I think I would expect the following output files:
- weighted_roc.tsv (which we always produce)
- test+mybed+het+snp_roc.tsv
- test+mybed+hom+snp_roc.tsv

I don't think I would expect to see:
- snp_roc.tsv
- het_roc.tsv
- hom_roc.tsv
- test_roc.tsv
- mybed_roc.tsv
- test+mybed_roc.tsv
- test+snp_roc.tsv
- test+het_roc.tsv
- test+hom_roc.tsv
- mybed+het+snp_roc.tsv
- mybed+hom+snp_roc.tsv
- mybed+het_roc.tsv
- mybed+hom_roc.tsv

Agree?

rickymagner commented:
It's possible for the first point we could deprecate the old flag but leave it in for now, and print a warning to use the two new flags instead going forward? 

That's a good point that the intersection is a special case. So that means the mode is completely controlled by the presence/lack of the `--roc-cross-join` flag then? 

I think there is a use case in flexibility there. I agree for the het/hom vs snp/indel, have `het_snp`, `hom_snp`, and `snp` could be excessive since the latter is already captured by the first two (and can be deduced algebraically from the first presumably). But in the case of restriction to a `bed` region, you might want both `mybed+snp` and `snp`. Now the latter is not possible to deduce from the former.

Perhaps there can be some flexibility in how the user wants these to be handled by actually allowing them to specify when the `ALL` filter should be added there by doing something like:
```
... --roc-cross-join --roc-subset snp --roc-regions "mybed=mybed.bed" --roc-regions ALL
```
which would produce `mybed+snp` and also `ALL+snp` = `snp` (and would be printed as the latter).

Now `ALL` would be a sort of keyword argument that means "also do the cross product where this value has no restriction added to the ROC computations." What do you think? Is it too confusing?

Lenbok commented:
OK, good discussion. Here's what I think we should do:

* For the flags, lets leave them as they are currently, and when I update the documentation to describe the new flag I'll talk about the notional categories within `--roc-subset` as being gt vs variant related.

* For the four per-category `RocFilter.ALL`s, I think that for now we should keep things simple by just adding them when the category is otherwise empty. It'll keep the number of output files managable and it's forward compatible with your suggestion of allowing special `ALL` command-line keyword (although we'd probably want to do the splitting out new flags for gt vs variant subsets at that time).

(BTW, are you aware of `rtg vcf2rocplot`, which makes it fairly cheap to generate new ROCs after the primary `rtg vcfeval` run as long as you output annotated VCFs?)

rickymagner commented:
Sounds good on the flags. For the `ALL`s, I think I agree that generally it probably doesn't make sense to always add it for each category (like the ones I mentioned before, e.g. Het/Hom). I suppose if I want to also get a "whole genome" default interval value, I can always input a whole genome bed file to cover that case. 

I wasn't aware of `vcf2rocplot`, but it sounds like that would be subject to the same issues of trying to combine multiple filters at once, e.g. getting `het+snp` wouldn't currently be possible for the same reasons `vcfeval` can't (which this PR addresses). It's possible that tool could get a similar upgrade, but is probably beyond the scope of this PR.

So it sounds like the only thing left to do on my end would be: change the way the filters are collected by only adding an `ALL` if the user provided no filters for each category, correct? I could probably get that done early next week and update the tests if needed so it'll be ready to merge.

Lenbok commented:
@rickymagner I just realized that as it stands this will actually break `rtg vcf2rocplot` and `rtg cnveval` (they both make use of the VcfEvalCli functions for roc filter setup). Obviously they don't have unit tests for this roc subsetting otherwise there'd be a test failure. Essentially `rtg cnveval` defines (only) two options for `--roc-subset`, `del` and `dup`. The new `VcfEvalCli.getRocFilters` will drop those guys on the floor. `rtg vcf2rocplot` is intended to allow processing either vcfeval or cnveval outputs, so it allows the union of the cnveval and vcfeval roc subsets (and simililarly the cnv roc subset processing will no longer work).

Can we add an extra vararg parameter to `getRocFilters`, like
```
  public static Set<RocFilter> getRocFilters(CFlags flags, List<RocFilter> fallbackRocFilters, Set<String>... rocSubsetGroups) throws IOException {
```
where vcfeval would call it passing in two two sets (the same as you've currently got internally as `genotypeOptions` and `variantOptions`), cnveval would call it with one set containing the del and dup filter names, and vcf2rocplot would call it with three sets. The innards of `getRocFilters` would need to handle the variable number of subset groups.

rickymagner commented:
Ah I didn't know `cnveval` exists as it's not in the RTG manual, but sure enough it has a `--help` output on the command line. I think your suggestion makes sense to add another input to the method so different tools can configure how to call it with `rocSubsetGroups` varying. Can you say a little more about what values `vcf2rocplot` should call the method with? I can try to make these changes hopefully sometime this week. Of course the "best" fix would be to probably make some sort of `RocProducingCliTool` abstract class both can inherit from instead of using the `vcfeval` static method to invoke this, but that's probably too serious a refactor for this PR.

Lenbok commented:
`cnveval` is a fairly early stage tool that we are planning to promote in the next release, which is why it doesn't appear in the manual or output of `rtg --help`. I agree regarding pulling our that shared functionality at some point in the future.

`vcf2rocplot` could either supply three different `rocSubsetGroups` (One for the `cnveval` subsets, and two corresponding to the `vcfeval` subsets for genotype and variant type), or alternatively you might consider the `cnveval` subsets `del` and `dup` as just being additional members of the variant type subset.

rickymagner commented:
Nice, good to know. Do you think it's still possible this PR would be able to get into a minor release before whenever you're thinking of that next major release including that tool? 

So to be clear for the refactoring of `getRocFilters`, I should:
* Add the new `rocSubsetGroups` input to the function, which would be a set containing elements like `snp, non-snp, indel, etc...` for `vcfeval` and `del, dup` for `cnveval`.  Do you want the method to also check consistency of these choices with the tool? If there are filters for `del` and `dup` which make sense as DEL/INS for `vcfeval`, it could be an addition of new options for the tool and handled the same way, no?
* Remove the addition of `RocFilter.ALL` to each category in the `roc-cross-join` mode unless no filtering conditions are given for that category.

Did I miss anything else? I think we might want a test or two for the `vcf2rocplot` and `cnveval`, right? Although I'm not adding the cross join functionality to them, so perhaps it doesn't apply?

Lenbok commented:
> Nice, good to know. Do you think it's still possible this PR would be able to get into a minor release before whenever you're thinking of that next major release including that tool?

I'm thinking we'd do it as one release, but we would try and do it promptly.

>     * Add the new `rocSubsetGroups` input to the function, which would be a set containing elements like `snp, non-snp, indel, etc...` for `vcfeval` and `del, dup` for `cnveval`.  Do you want the method to also check consistency of these choices with the tool? If there are filters for `del` and `dup` which make sense as DEL/INS for `vcfeval`, it could be an addition of new options for the tool and handled the same way, no?

I think a sufficient consistency check is that when filtering through the user-selected roc subsets, they should always fall into one of the `rocSubsetGroups`, and raise `NoTalkbackSlimException` otherwise. I wouldn't change the set of subset options currently available to the tools.

>     * Remove the addition of `RocFilter.ALL` to each category in the `roc-cross-join` mode unless no filtering conditions are given for that category.

Correct.

> Did I miss anything else? I think we might want a test or two for the `vcf2rocplot` and `cnveval`, right? Although I'm not adding the cross join functionality to them, so perhaps it doesn't apply?

By making it part of `getRocRilters` you **have** added that functionality to them :-).

You can probably just update/copy the existing tests in Vcf2RocplotTest and CnvEvalTest for these to check for expected output files, and possibly again when adding `--roc-cross-join`.

rickymagner commented:
Hi @Lenbok, I took a stab at refactoring `getRocFilters` to follow the scheme outlined above. I updated the vcfeval test and it was all passing on my machine. Let me know what you think of the latest commit when you get a chance. I might've misunderstood something, but did you also want me to change anything in the cnveval / roc2plot tests? Or would those still be fine if the old tests are passing?

On a side note, I added the `NoTalkbackSlimException` in the case the `getRocFilters` groups doesn't fit the set for the tool. I was curious about what the "talkback" stuff is in the tool kit. It seems like a way to report errors automatically to a database somewhere. Is there a global flag to shut this off by default? I was worried when experimenting earlier that I might've been spamming some database with errors that were just me debugging some changes. I also wonder if some people might be nervous about sending sensitive filenames/paths in the log unknowingly if using the tools.

Lenbok committed:
Make cnveval test check rocs

Lenbok commented:
Hi @rickymagner I added a couple more commits to this PR, take a look.

I also tweaked the current `cnveval` unit test to use roc subsets, and they continue to pass. I also experimented with adding the `--roc-cross-join` flag to that unit test, it has the effect of enabling rescaling on the del/dup roc files. (In this PR the CombinedRocFilter has rescaling enabled when there were no `--roc-expr` filters). We might instead want to set rescale to null, so that it uses the global default, but I'm curious what you expect the behaviour should be when using cross join mode.

Regarding the crash reporting, the setting for that is in the `rtg.cfg` file that gets created when you run `rtg` for the first time. Crash reporting is opt-out (there's a separate opt-in usage logger that excludes command lines for seeing how often the different subcommands are used).

rickymagner commented:
Hi Len, thanks for the extra commits. I think it's a good idea to refactor to make the combined roc filters separately, and thanks for adding some extra docs. For the rescaling: I think I just did whatever seemed to "make sense" based on some experimenting, or maybe an error was thrown when the expr filters didn't have it set. I don't have any expectations on what "should" happen here necessarily, so would defer to whatever you think is appropriate. IIRC, I think it only has a small effect on the outputs, right?

For the reporting, now that I think about it I don't think I've ever had an issue with the program hanging in a docker image. I suppose a default cfg file must be used in that case where you aren't running interactively?

Lenbok commented:
Hi Ricky,

I've adjusted the logic for whether to enable rescaling. Now it requires all the sub-filters to have rescaling enabled in order to turn on. I think this better matches the principle of least surprise, in that using `--roc-cross-join` when there is only a single "dimension" will not alter the use of rescaling. If you're using multiple dimensions it'll turn off rescaling if any of them have rescaling turned of, and the rocs will then have raw numbers.

I think this looks OK now, so if you're good with it, I'll do a little commit squashing and merge.

rickymagner commented:
Great, that sounds good to me. I think this is good to go for me, so whenever you're ready to merge that would be great! Thanks for all your help in getting this to work. Looking forward to the tagged release hopefully coming soon.

Lenbok commented:
BTW, thanks for the contribution! We'll work on getting a new release out.

rickymagner commented:
Thanks, and thanks for all your help! Looking forward to the formal release to incorporate into our pipeline.
